src/Confluent.Kafka/Consumer.cs

```diff
@@ -695,50 +733,42 @@ protected virtual void Dispose(bool disposing)
 
 
         /// <summary>
-        ///     Set the partitions assigned handler.
+        ///     Set the consumer group rebalance handler. Except for the initial partition assignment
+        ///     and final assignment revocation, a group rebalance will cause this handler to be
+        ///     called twice. The first time to advise that the current assignment has been revoked
+        ///     and the second time to advise of the new assignment.
         /// 
-        ///     If you do not call the <see cref="Confluent.Kafka.Consumer{TKey,TValue}.Assign(IEnumerable{TopicPartition})" />
-        ///     method (or another overload of this method) in this handler, or do not specify a partitions assigned handler,
-        ///     the consumer will be automatically assigned to the partition assignment set provided by the consumer group and
-        ///     consumption will resume from the last committed offset for each partition, or if there is no committed offset,
-        ///     in accordance with the `auto.offset.reset` configuration property. This default behavior will not occur if
-        ///     you call Assign yourself in the handler. The set of partitions you assign to is not required to match the
-        ///     assignment provided by the consumer group, but typically will.
+        ///     If you do not call an Assign or Unassign method in this handler, or do not specify
+        ///     a handler, partitions will be assigned (or unassigned) automatically, matching the
+        ///     assignment dictated by the consumer group. This default behavior will not occur if
+        ///     you call Assign yourself in the handler. The set of partitions you assign to is not
+        ///     required to match the assignment provided by the consumer group, but typically will.
+        /// 
+        ///     Consumption will resume from the last committed offset for each partition (unless an
+        ///     offset is manually specified in a call to Assign), or if there is no committed
+        ///     offset, in accordance with the `auto.offset.reset` configuration property.
         /// </summary>
         /// <remarks>
         ///     Executes as a side-effect of the Consumer.Consume call (on the same thread).
         /// </remarks>
-        public void SetPartitionsAssignedHandler(Action<IConsumer<TKey, TValue>, List<TopicPartition>> value)
+        public void SetRebalanceHandler(Action<IConsumer<TKey, TValue>, RebalanceEvent> value)
         {
             if (value == null)
             {
-                partitionsAssignedHandler = null;
-                return;
+                throw new ArgumentNullException("Rebalance handler cannot be null.");
             }
-            partitionsAssignedHandler = partitions => value(this, partitions);
-        }
 
+            if (this.rebalanceHandler != null)
```

Comment:
locking

Reply:
yeah. 

As an aside, I'm still questioning the decision to have some of the set methods on the consumer - finding the tradeoffs tricky to assess. If this changes though, it'll be a separate PR.

Reply:
So you're thinking of moving the rebalancehandler to the builder?

Reply:
I am, but I want to work through it more and defer that to another PR. it also affects the offsets committed handler.